src/StorageRegistry.sol

```diff
@@ -0,0 +1,321 @@
+
+        rentedUnits += units;
+        emit Rent(msg.sender, fid, units);
+    }
+
+    /**
+     * @notice Rent storage for multiple fids. The caller must provide an exact payment amount equal to
+     *         the sum of the prices for each fid's storage allocation.
+     *
+     * @param fids  An array of fids.
+     * @param units An array of storage unit quantities. Must be the same length as the fids array.
+     */
+    function batchRent(uint256[] calldata fids, uint256[] calldata units) external payable whenNotDeprecated {
+        if (fids.length == 0 || units.length == 0) revert InvalidBatchInput();
+        if (fids.length != units.length) revert InvalidBatchInput();
+
+        uint256 _ethUsdPrice = ethUsdPrice();
+        uint256 _usdUnitPrice = usdUnitPrice;
+
+        uint256 totalCost;
+        for (uint256 i; i < fids.length; ++i) {
+            uint256 qty = units[i];
+            if (qty == 0) continue;
+            if (rentedUnits + qty > maxUnits) revert ExceedsCapacity();
+            totalCost += _price(qty, _usdUnitPrice, _ethUsdPrice);
+            rentedUnits += qty;
+            emit Rent(msg.sender, fids[i], qty);
+        }
+
+        if (msg.value != totalCost) revert InvalidPayment();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              PRICE VIEWS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Cost in wei to rent one storage unit.
+     */
+    function unitPrice() public view returns (uint256) {
+        return price(1);
+    }
+
+    /**
+     * @notice Calculate the cost in wei to rent the given number of storage units.
+     *
+     * @param units Number of storage units.
+     */
+    function price(uint256 units) public view returns (uint256) {
+        return _price(units, usdUnitPrice, ethUsdPrice());
+    }
+
+    function ethUsdPrice() public view returns (uint256) {
+        /* Ensure that the L2 sequencer is up. */
+        (, int256 sequencerUp, uint256 startedAt,,) = uptimeFeed.latestRoundData();
+        if (sequencerUp != 0) revert SequencerDown();
+
+        /* If the L2 sequencer recently restarted, ensure the grace period has elapsed. */
+        uint256 timeSinceUp = block.timestamp - startedAt;
+        if (timeSinceUp < L2_DOWNTIME_GRACE_PERIOD) revert GracePeriodNotOver();
+
+        /* Get and validate the Chainlink ETH/USD price. */
+        (uint80 roundId, int256 answer,, uint256 updatedAt, uint80 answeredInRound) = priceFeed.latestRoundData();
+        if (answer <= 0) revert InvalidPrice();
+        if (updatedAt == 0) revert IncompleteRound();
+        if (answeredInRound < roundId) revert StalePrice();
+
+        return uint256(answer);
+    }
+
+    function _price(uint256 units, uint256 usdPerUnit, uint256 ethPerUsd) internal pure returns (uint256) {
+        return units * usdPerUnit * 1e18 / ethPerUsd;
```

Comment:
should we be protecting against overflows here if the `units` value is set to something really large in the batch method?

Reply:
Checked math should keep this multiplication and the accumulator in `batchRent` safe against overflow, but we might revert early with a checked math error during the price calculation in `batchRent` if the caller passes a very high `units` value. The error is not super friendly, but it's safe. This seems OK to me, what do you think?

Reply:
that seems reasonable, i forget math is default checked now